### PR TITLE
8267353: java/net/SctpSanity.java fails due to Protocol not supported

### PR DESCRIPTION
**Pull request by DamonFool**
Hi all,

java/net/SctpSanity.java fails on some of our test machines due to Protocol not supported.
The reason is that the test fails to detect all the cases when a machine doesn't support SCTP.

The fix just follows what are done in [1][2][3] to detect unsupported paltforms at the beginning.

Thanks.
Best regards,
Jie

[1] https://github.com/openjdk/jdk/blob/master/test/jdk/com/sun/nio/sctp/SctpMultiChannel/Util.java#L59
[2] https://github.com/openjdk/jdk/blob/master/test/jdk/com/sun/nio/sctp/SctpServerChannel/Util.java#L59
[3] https://github.com/openjdk/jdk/blob/master/test/jdk/com/sun/nio/sctp/SctpChannel/Util.java#L59

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Issue
 * [JDK-8267353](https://bugs.openjdk.java.net/browse/JDK-8267353): java/net/SctpSanity.java fails due to Protocol not supported

### Contributors
 * Junji Wang `<junjiwang@tencent.com>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4199/head:pull/4199` \
`$ git checkout pull/4199`

Update a local copy of the PR: \
`$ git checkout pull/4199` \
`$ git pull https://git.openjdk.java.net/jdk pull/4199/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4199`

View PR using the GUI difftool: \
`$ git pr show -t 4199`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4199.diff">https://git.openjdk.java.net/jdk/pull/4199.diff</a>

</details>

### CODE REVIEW

**Comment by ChrisHegarty** [comment]
This is probably ok, but a little surprising. Can you say a more about the machine setup?

**Comment by dfuch** [comment]
I am intrigued - which exception do you get on the machine where SCTP is not supported?
SctpChannel.open() is supposed to throw `UnsupportedOperationException` - If the SCTP protocol is not supported, and the test catches  `UnsupportedOperationException`.

**Comment by DamonFool** [comment]
> I am intrigued - which exception do you get on the machine where SCTP is not supported?

Hi @dfuch , we got this exception, which was copied from the JBS.

```
java.net.SocketException: Protocol not supported
        at jdk.sctp/sun.nio.ch.sctp.SctpNet.socket0(Native Method)
        at jdk.sctp/sun.nio.ch.sctp.SctpNet.socket(SctpNet.java:81)
        at jdk.sctp/sun.nio.ch.sctp.SctpChannelImpl.<init>(SctpChannelImpl.java:138)
        at jdk.sctp/com.sun.nio.sctp.SctpChannel.open(SctpChannel.java:170)
        at SctpSanity.testSctpChannel(SctpSanity.java:58)
        at SctpSanity.main(SctpSanity.java:50)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
        at java.base/java.lang.Thread.run(Thread.java:831)
```

**Comment by DamonFool** [comment]
> This is probably ok, but a little surprising. Can you say a more about the machine setup?

What do you mean by `setup`?
And how can I get these setups?
Thanks.

**Comment by AlanBateman** [comment]
> java.net.SocketException: Protocol not supported
>         at jdk.sctp/sun.nio.ch.sctp.SctpNet.socket0(Native Method)

Do we know if the socket(IPPROTO_SCTP) or setsockopt(IPPROTO_SCTP/SCTP_EVENTS) is failing? I would have expected the former to check for the not supported case (and the latter to close the fd, but that is a separate issue).

**Comment by ChrisHegarty** [comment]
> > This is probably ok, but a little surprising. Can you say a more about the machine setup?
> 
> What do you mean by `setup`?
> And how can I get these setups?
> Thanks.

Let me rephrase.

You said: "The reason is that the test fails to detect all the cases when a machine doesn't support SCTP."

My question is: Please explain the case (or set of cases) where your "machine doesn't support SCTP" ?

**Comment by DamonFool** [comment]
> Do we know if the socket(IPPROTO_SCTP) or setsockopt(IPPROTO_SCTP/SCTP_EVENTS) is failing?

How can I get to know if the socket(IPPROTO_SCTP) or setsockopt(IPPROTO_SCTP/SCTP_EVENTS) is failing?
Any suggestions?
Thanks.

**Comment by DamonFool** [comment]
java/net/SctpSanity.java is the only one, which failed on the machine.
Other sctp related tests like jdk/com/sun/nio/sctp/SctpChannel/Connect.java are all fine.

**Comment by AlanBateman** [comment]
> How can I get to know if the socket(IPPROTO_SCTP) or setsockopt(IPPROTO_SCTP/SCTP_EVENTS) is failing?

`strace -f` should help, or just add trace message. If socket(IPPROTO_SCTP) is failing then it suggests to me that we are missing the check for ENOPROTO.

**Comment by DamonFool** [comment]
> `strace -f` should help, or just add trace message.

Thanks for your help @AlanBateman .
I got the following msg (Connection timed out).
```
[pid 93582] socket(PF_INET, SOCK_STREAM, IPPROTO_SCTP <unfinished ...>
[pid 93557] <... futex resumed> )       = -1 ETIMEDOUT (Connection timed out)
[pid 93557] futex(0x2b05c03cff78, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 93557] getrusage(0x1 /* RUSAGE_??? */, {ru_utime={0, 0}, ru_stime={0, 0}, ...}) = 0
[pid 93557] getrusage(0x1 /* RUSAGE_??? */, {ru_utime={0, 0}, ru_stime={0, 0}, ...}) = 0
[pid 93557] getrusage(0x1 /* RUSAGE_??? */, {ru_utime={0, 0}, ru_stime={0, 0}, ...}) = 0
[pid 93557] getrusage(0x1 /* RUSAGE_??? */, {ru_utime={0, 0}, ru_stime={0, 0}, ...}) = 0
[pid 93557] futex(0x2b05c03cffa4, FUTEX_WAIT_BITSET_PRIVATE, 5, {10828639, 700862386}, ffffffff <unfinished ...>
[pid 93582] <... socket resumed> )      = -1 EPROTONOSUPPORT (Protocol not supported)
```
Thanks.

**Comment by AlanBateman** [comment]
Okay, so I think the issue is that EPROTONOSUPPORT is not handled as I would have expected that to cause UOE to be thrown. I suspect the isSCTPSupported check in Util.java (used by the tests in jdk/test/com/sun/nio/sctp) are masking this. My memory is that there were several relilability issues with SCTP in the early days and it might be that this is why they skip when any error is returned creating the SCTP socket.

**Comment by ChrisHegarty** [comment]
There is no check for ENOPROTO in the JDK SCTP code. There is an implicit assumption that SCTP is "supported" when the native libsctp.so.1 library successfully loads and can be introspected. Maybe there should be an explicit ENOPROTO in Java_sun_nio_ch_sctp_SctpNet_socket0.

Maybe the machine in question has the libsctp.so.1 library, but the kernel module is not loaded or something?

```
$ which checksctp
/usr/bin/checksctp

$ /usr/bin/checksctp
SCTP supported

$ lsmod | grep sctp
sctp                  299008  2 
libcrc32c              16384  4 nf_conntrack,nf_nat,xfs,sctp

$ modinfo sctp
filename:       /lib/modules/4.14.35-1902.0.18.el7uek.x86_64/kernel/net/sctp/sctp.ko.xz
license:        GPL
description:    Support for the SCTP protocol (RFC2960)
author:         Linux Kernel SCTP developers <linux-sctp@vger.kernel.org>
alias:          net-pf-10-proto-132
alias:          net-pf-2-proto-132
srcversion:     CA88E7C1DCACA58B08A8E35
depends:        libcrc32c
retpoline:      Y
intree:         Y
name:           sctp
vermagic:       4.14.35-1902.0.18.el7uek.x86_64 SMP mod_unload modversions 
sig_id:         ...

$ /sbin/modprobe -n -v sctp
```

**Comment by msheppar** [comment]
if sctp is configured on a linux system would there exist a  /proc/net/sctp  and have an entry in /proc/net/protocols ?

**Comment by DamonFool** [comment]
> if sctp is configured on a linux system would there exist a /proc/net/sctp and have an entry in /proc/net/protocols ?

No /proc/net/sctp and no entry in /proc/net/protocols.
But I don't think it's the reason since it passed on another machine without /proc/net/sctp and the entry too.
Thanks.

**Comment by DamonFool** [comment]
> Maybe the machine in question has the libsctp.so.1 library, but the kernel module is not loaded or something?

Yes, you are right.
We do have libsctp.so.1 and no kernel module.

So what's your suggestion next?
Thanks.

**Comment by DamonFool** [comment]
> > if sctp is configured on a linux system would there exist a /proc/net/sctp and have an entry in /proc/net/protocols ?
> 
> No /proc/net/sctp and no entry in /proc/net/protocols.
> But I don't think it's the reason since it passed on another machine without /proc/net/sctp and the entry too.
> Thanks.

After `sudo yum install lksctp-tools`, which would install libsctp.so.1, both the machines fail now.
So @ChrisHegarty is right: the machine in question has the libsctp.so.1 library, but the kernel module is not loaded.
Thanks.

**Comment by ChrisHegarty** [comment]
@DamonFool Thanks for checking this, and helping track down the root cause of the issue.

The machine configuration is a little unusual (since it has the user-level libraries installed, while the kernel module is not loaded). Clearly, to workaround the issue, the "sctp" kernel module could be loaded, but that will just serve to mask/hide the underlying problem.

The bug is really in the Java_sun_nio_ch_sctp_SctpNet_socket0 native method ( SctpNet.c ) - there should be an explicit check for ENOPROTO after calling socket(2), throwing UnsupportedOperationException if that is the case.

@DamonFool  I would much prefer to keep the test as it is and fix the underlying bug, but this may be more than you want to do? If so, I'll get this done myself.

**Comment by DamonFool** [comment]
> @DamonFool I would much prefer to keep the test as it is and fix the underlying bug, but this may be more than you want to do? If so, I'll get this done myself.

Please feel free to do so.
I'll close this pr soon.
Thanks.
